tree_unique_args/src/loop_strength_reduction.egg

```diff
@@ -0,0 +1,111 @@
+(ruleset loop-strength-reduction)
+
+;; Change multiplication of loop variable to addition to an accumulator
+;;
+;; ORIGINAL
+;; a = 0
+;; c = 3
+;; for  i = 0 to n:
+;;     a = i * c
+;;
+;; OPTIMIZED
+;; a = 0
+;; c = 3
+;; d = 0
+;; for i = 0 to n:
+;;     a = d
+;;     d += c
+
+; Finds invariants/constants within a loop.
+; Columns: loop; value of invariant in inputs; value of invariant in outputs
+; TODO: once everything else is working, change this to just use an ID to ref the loop
+(relation Inv (Expr Expr Expr))
+(rule (
+    (= loop (Loop id (All (Parallel) inputs) (All (Sequential) (Pair pred (All (Parallel) outputs)))))
+    (= (ListExpr-ith outputs i) (Get (Arg id) i)))
+    ((Inv loop (ListExpr-ith inputs i) (Get (Arg id) i))) :ruleset always-run)
+
+; NOTE: I removed this rule for now. This is because it is not correct to put
+; (Num id n) into the inputs of this loop. It has the id of the loop, so it
+; belongs within the context of the outputs, and therefore relies on RefIdOf.
+; TODO: handle this and add corresponding test case
+; This note I copied from the previous encoding (unique args) is actually incorrect:
+    ;; Note: with the milestone encoding, we had to use
+    ;; Body-contains-Operand here. Here, we already know
+    ;; that c is within loop because it has the correct id. 
+(rule (
+    (= c (Num id n))
+    (= loop (Loop id inputs predoutputs))
+    (ExprHasRefId loop outer-id))
```

Comment:
Indeed, with newly merged #291, you can get the id of outer region of `loop` by inspecting the id to inputs.